src/lib/cache/abstract_cache.hpp

```diff
@@ -1,52 +1,36 @@
 #pragma once
 
+#include <atomic>
+#include <optional>
+#include <unordered_map>
 #include <utility>
 
-#include <boost/iterator/iterator_facade.hpp>
-
 namespace opossum {
 
+inline constexpr size_t DEFAULT_CACHE_CAPACITY = 1024;
+
 // Generic template for a cache implementation.
+// It guarantees the capacity to be thread-safe. Any other guarantees have to be fulfilled by the implementation.
 template <typename Key, typename Value>
-class AbstractCacheImpl {
+class AbstractCache {
  public:
-  using KeyValuePair = typename std::pair<Key, Value>;
-
-  struct AbstractIterator {
-    virtual ~AbstractIterator() = default;
-
-    virtual void increment() = 0;
-    virtual bool equal(const AbstractIterator& other) const = 0;
-    virtual const KeyValuePair& dereference() const = 0;
-  };
-
-  class ErasedIterator
-      : public boost::iterator_facade<ErasedIterator, KeyValuePair const, boost::forward_traversal_tag> {
-   public:
-    explicit ErasedIterator(std::unique_ptr<AbstractIterator> it) : _it(std::move(it)) {}
-
-   private:
-    friend class boost::iterator_core_access;
-
-    void increment() { _it->increment(); }
-    bool equal(const ErasedIterator& other) const { return _it->equal(*other._it); }
-    const KeyValuePair& dereference() const { return _it->dereference(); }
+  explicit AbstractCache(size_t capacity = DEFAULT_CACHE_CAPACITY) : _capacity(capacity) {}
 
-    std::unique_ptr<AbstractIterator> _it;
-  };
-
-  explicit AbstractCacheImpl(size_t capacity) : _capacity(capacity) {}
-
-  virtual ~AbstractCacheImpl() {}
+  virtual ~AbstractCache() {}
 
   // Cache the value at the given key.
   // If the new size exceeds the capacity an item will be evicted.
   // Depending on the underlying strategy, the parameters for cost and size may be used.
   // If they are not intended to be used, we specify constant default values here.
   virtual void set(const Key& key, const Value& value, double cost = 1.0, double size = 1.0) = 0;
 
+  // Tries to fetch the cache entry for the query into the result object. Returns true if the entry was found, false
+  // otherwise.
+  virtual std::optional<Value> try_get(const Key& key) = 0;
+
   // Get the cached value at the given key.
   // Causes undefined behavior if the item is not in the cache.
+  // This cannot be thread-safe since we return a reference here.
```

Comment:
Is this actually used somewhere outside of the tests or can we get rid of it? In that case, I would add a comment to try_get, explaining why it does not return a reference.

Reply:
After taking a quick look, it is only used in tests. I'll remove `get()`.